rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,438 @@
+
+This would require Tendermint to support all prior block protocol version. This
+would be done by somehow finding what the version of the structure is and
+processing it accordingly. This mimics a similar design model to the Cosmos SDK
+where all legacy code is stored in the repository and conditionals are used to
+indicate how messages are processed.
+
+One could imagine a directory structure for each module as
+such:
+
+```go
+consensus // module
+|---- state.go
+|---- state_test.go
+|       ...
+|---- votes.go // current version (v10)
+|       ...
+|---- legacy
+   |---- v8 // prior supported versions
+      |---- votes.go
+      |      ...    // other data structures ( and pure functions )
+   |---- v9
+      |---- votes.go
+      |      ...
+```
+
+The module, depending on the types, may statically cast each of the different
+versions or define an interface that it expects all versions to follow. For
+example:
+
+```go
+func (blockExec *BlockExecutor) CreateProposalBlock(state State) (BlockI, error)
+{
+ switch state.Version.Block {
+ case 9:
+  return v9.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ case 10:
+  return v10.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ default:
+  return fmt.Errorf("block version %d not recognised", state.Version.Block)
+ }
+}
+```
+
+In this case, Block which falls under a data structure of the block protocol
+could have many versions and is represented here as an interface. `State` and
+`BlockExecutor` are structs outside of the block protocol so would be
+internalized to allow for changes (so long as this wouldn't change the overall
+behavior).
+
+### Current Use of Versioning
+
+Records of versions are currently stored in `State` and in the `Header` of each
+block. This already covers a broad base of scenarios. The consensus reactor can
+use the version from state to ensure it reads `Vote`s and `Partsetheader`s
+correctly. Similarly, fast sync, the block executor and evidence module, whom
+all interact with block protocol data structures also all have access to state.
+The next few sections will explore the aspects of Tendermint that will need to
+be modified to prepare for multi versioning support.
+
+### RPC
+
+Supporting soft upgrades means that the RPC will need to be able to deliver
+information across different versions. The RPC will therefore need to know which
+heights corresponds to which version and then be able to communicate this to
+the clients. This would most likely require clients to be versatile to handling
+the different block versions. For example block explorers would also take on the
+burden of supporting all block versions across a chain.
+
+### Light Clients
+
+Similar to RPC, light clients will also need to be wary of block protocol
+changes and be able to handle them accordingly. Depending on these changes,
+there might be the need for special verification logic across signed headers
+with different versions.
```

Comment:
Similar as above, light client will need to encode versions for height ranges (for example v10 up to height 1000, v11 up to height 2500, v12 above 2500), and then depending on the height for which query is made, right version should be used.

Reply:
yup. Although the version is already encoded in the header. The light client just needs to read this once it is requested and then process the header accordingly.